### PR TITLE
Initial charm

### PR DESCRIPTION
**Pull request by marceloneppel**
This PR introduces the following implementations:

* Initial charm structure based on the [operator template](https://github.com/canonical/operator-template).
* PostgreSQL deployment on bionic and focal series (without HA at this moment) with superuser password sharing among the units.
* Unit and integration tests.

### CODE REVIEW

File: src/charm.py
Comment (jnsgruk) [inline src/charm.py:104]: might be worth using `distro`

```python
import distro
distro.codename() # => 'focal'
```

File: src/cluster.py
Comment (jnsgruk) [inline src/cluster.py:81]: Might be worth using the `tempfile` module here

File: src/cluster.py
Comment (jnsgruk) [inline src/cluster.py:92]: this should be `logger.debug`

File: src/cluster.py
Comment (jnsgruk) [inline src/cluster.py:111]: Just a stylistic point, and by no means mandatory, but this could probably be quite cleanly represented as a lambda when you call filter?

File: src/cluster.py
Comment (jnsgruk) [inline src/cluster.py:119]: Probably worth documenting the args in this docstring

File: src/cluster.py
Comment (jnsgruk) [inline src/cluster.py:223]: `logger.debug` for this :)

File: tests/integration/test_focal.py
Comment (jnsgruk) [inline tests/integration/test_focal.py:1]: I think this file could be dropped and you could use pytest's parameterize feature in the other integration test to just deploy for both series'

File: tests/integration/test_bionic.py
Comment (jnsgruk) [inline tests/integration/test_bionic.py:57]: Could we add another test here that gets cluster status/number of members/etc?

Comment (jnsgruk) [reply tests/integration/test_bionic.py:57]: On reflection, I wonder if its possible to use pylibjuju to pull files from the unit and check their contents are what we expect (for example config files?)

**Review by jnsgruk — CHANGES_REQUESTED** [review]
Overall this is a great start! Just a few comments on tests and a couple of nits.

File: src/cluster.py
Comment (wolsen) [inline src/cluster.py:101]: Just an FYI - not sure if you are aware of [py-postgresql](https://pypi.org/project/py-postgresql/). While it also has database drivers to query and talk with the database, it has some built-ins regarding [cluster management](https://pythonhosted.org/py-postgresql/cluster.html) which may be more convenient; however I think it does raw process management rather than leveraging systemd (though I'm sure that could be subclassed and extended that way, and even possibly contributed back upstream).

The advantage of using this library is that there's already users who are using it and helping to work out some various bugs on connections, cluster status, etc, etc.

File: src/cluster.py
Comment (wolsen) [inline src/cluster.py:136]: Is it intended that these will change at some point? Since they're all constants currently, doesn't seem like much of a need to provide them as variables to the template itself rather than just adding the contents to the file directly.

File: templates/pg_hba.conf.j2
Comment (wolsen) [inline templates/pg_hba.conf.j2:8]: nit: If a developer includes a newline in the comment this will break. I suppose its dev-mode and should be caught, but it might cause unintended confusion later.

File: tox.ini
Comment (wolsen) [inline tox.ini:37]: nit: could just throw these in the test-requirements.txt rather than inline some and not others. Its kind of nice if there's generally one or two places to go and its consistent. Gets to be an awkward contributor experience if you have really dive into the different ways dependencies are managed and defined.

**Review by wolsen — COMMENTED** [review]
Overall, looks like a really solid start on this. A few comments inline with some initial thoughts.

File: src/charm.py
Comment (wolsen) [inline src/charm.py:49]: You essentially have a factory pattern here, but I think your factory is in the wrong place. Rather than include the `_get_postgresql_version` in the charm itself, it might make more sense to have a getPostgresqlCluster() method which introspects aspects of the system and returns the correct PostgresqlCluster object.

The only real reason you have the get_postgresql_version is for version strings in file paths, which could probably be contained within the PostgresqlCluster class itself; removing the need for a factory altogether.

File: src/charm.py
Comment (wolsen) [inline src/charm.py:116]: See my above comment about factory...

However, here you're hardcoding which versions map to which release; which works but its a bit static. You'll have to come in and update the mapping for each release.

An alternative might be to determine which postgresql version will be installed by referencing the apt-cache and pulling the version string from there. Then it becomes sufficiently insulated from the distro packaging tasks.

**Review by wolsen — CHANGES_REQUESTED** [review]

File: actions.yaml
Comment (jnsgruk) [reply actions.yaml:5]: I've gone with 'get-initial-password' on other charms as a nod that it's generated, and should probably be changed. Until secrets arrives in Juju, I've not been providing a way to retrieve passwords set by the user.

File: tox.ini
Comment (jnsgruk) [reply tox.ini:37]: Just a comment here: in general I've been encouraging charm teams to have all runtime dependencies in requirements.txt, and all dev/test dependencies in the respective environment in tox.ini - but either way if test-requirements.txt is knocking around we could kill it in favour of this.

File: src/charm.py
Comment (marceloneppel) [reply src/charm.py]: Good to know. I changed it to use the `distro` package.

File: src/cluster.py
Comment (marceloneppel) [reply src/cluster.py]: Fixed.

File: src/cluster.py
Comment (marceloneppel) [reply src/cluster.py]: Fixed.

File: src/cluster.py
Comment (marceloneppel) [reply src/cluster.py]: Yes, you are right. I removed this function and used a lambda expression.

File: src/cluster.py
Comment (marceloneppel) [reply src/cluster.py]: It makes sense. I've documented the args of this method.

File: tests/integration/test_focal.py
Comment (marceloneppel) [reply tests/integration/test_focal.py]: Great suggestion Jon. I changed the code to use `pytest's parametrize` and also moved the charm `build` call to a `fixture` in order to run it just once (as we don't need different builds at this time for the two different series).

File: tests/integration/test_bionic.py
Comment (marceloneppel) [reply tests/integration/test_bionic.py]: I added a test to check the cluster status and a TODO to get the number of members in the replication when the replication is implemented.

Comment (marceloneppel) [reply tests/integration/test_bionic.py]: I also added a test to pull the configuration files contents and check that they are as expected.

File: actions.yaml
Comment (marceloneppel) [reply actions.yaml:5]: I makes sense, Billy. I'll change that.

File: src/charm.py
Comment (marceloneppel) [reply src/charm.py:50]: I agree, Billy! Thanks for catching that. I also use the version in the `apt.add_package` call (which didn't worked with the package `postgresql`, without the version - I'll check the call to make sure it's correct), which can also be retrieved with your suggestion about `apt-cache` in the comment below. I'll change that to get the version in PostgresqlCluster object, which makes more sense.

File: src/charm.py
Comment (marceloneppel) [reply src/charm.py:113]: Thanks for the suggestion. I'll check `apt-cache` for the version info.

File: src/cluster.py
Comment (marceloneppel) [reply src/cluster.py:143]: Probably not. Accesses by relation will be made using the last rule. Therefore, the current file content is enough. I'll change to a file with all the content to copy to the file in the unit instead of the template.

File: templates/pg_hba.conf.j2
Comment (marceloneppel) [reply templates/pg_hba.conf.j2:8]: I agree.
This template will be removed as said in the comment above  (there will be no future changes to the configuration file).

File: src/charm.py
Comment (marceloneppel) [reply src/charm.py]: The version is now retrieved from installed packages or apt-cache using the apt library.

File: src/charm.py
Comment (marceloneppel) [reply src/charm.py]: I removed the factory. I also updated the apt library in order to install the postgresql package instead of the ones with the version in the name.

File: tox.ini
Comment (marceloneppel) [reply tox.ini:37]: I  kept the dev dependencies in tox.ini as we talked in the meetings today.

File: src/cluster.py
Comment (marceloneppel) [reply src/cluster.py]: I removed the template and created a file with all the contents to be copied to the server instance.

File: src/cluster.py
Comment (marceloneppel) [reply src/cluster.py:120]: Great Billy. I agree with what you wrote and it's really good to know about this package. I tested it but found some problems for our use case. For the first one, related to how Postgres version is displayed on Ubuntu, I submitted a [PR](https://github.com/python-postgres/fe/pull/116), which needs the approval of the repo maintainers. The second problem is related to run the library code as postgres user instead of root, which can be achieved by using su like you said in our meeting. As we talked, we can address this change in a future PR.

File: .github/workflows/ci.yaml
Comment (marceloneppel) [reply .github/workflows/ci.yaml:1]: Great. I added it to the github actions configuration file. I'm not sure, but I believe we may need to allow access from codecov to this repo.

File: actions.yaml
Comment (marceloneppel) [reply actions.yaml]: I changed the action to get-initial-passwor. I'll also change in the k8s charm.

File: src/charm.py
Comment (marceloneppel) [reply src/charm.py]: I removed this now as it's not needed anymore after the [update](https://github.com/canonical/operator-libs-linux/pull/35#issuecomment-993278821) in the apt library  and because the PostgresqlCluster class is now using the version retrieved through the apt library.

File: tests/integration/test_focal.py
Comment (marceloneppel) [reply tests/integration/test_focal.py]: I also moved the expected content of the configuration files from the code to files in the repo in order to not duplicate them, as they are used in multiple source and test files.

File: tox.ini
Comment (wolsen) [reply tox.ini:37]: I think this is reasonable and really up to the development team to make this decision. As a note, the openstack charms (of which I have most familiarity) do it by dropping requirements into various {test-,build-,}requirements.txt files and we occasionally sync centralized copies of these files across all charms to have standard dependencies across all openstack charm projects.

**Review by jnsgruk — APPROVED** [review]
LGTM Marcelo, thanks!

File: src/cluster.py
Comment (wolsen) [reply src/cluster.py:120]: That's unfortunate, but yeah - its not something required here anyways. I was beginning to wonder about k8s usage as well.

**Review by wolsen — APPROVED** [review]

**Pull request merged**
